Issue #2211: OneStatementPerLine has false match with try-with-resources

jonmbake opened the pull request:
`OneStatementPerLine#isMultilineStatement` was returning true when the `previousSibling` was a RESOURCES token.  The solution was to return false if  `previousSibling` is a RESOURCES token.

Let me know if there are any styling issues.

romani commented:
thanks for contribution

1)
> prevSibling.getType() != TokenTypes.RESOURCES

please move this out of `isMultilineStatement` , Resource is multiline but we ignore it by special request. So `isMultilineStatement` is correctly return `true` on it.

2)
Please add UT cases 
```
try( AutoCloseable i = new java.io.StringReader("foo");
      AutoCloseable k = new java.io.StringReader("foo");) {
        } catch (Exception e1) {
        }

try( AutoCloseable i = new java.io.StringReader("foo");AutoCloseable k = new java.io.StringReader("foo");) {
        } catch (Exception e1) {
        }
```

3)
All functional changes should pass functional testing of real sources to avoid regressions.
This process is mostly automated. Please read README of https://github.com/checkstyle/contribution/tree/master/checkstyle-tester . We need "DIFF-REPORT" of Check before fix and after fix. Diff report need to shared with me for verification.

I appreciate your help, you are ok to ask any questions to let you finish. You can join a gitter.im chat room to get help from real people.

jonmbake commented:
@romani Thanks for your feedback.

I made the update and ran the regression tests with the checkstyle-tester.  When I attempt to user [ahsm](https://github.com/attatrol/ahsm) to generate the DIFF-REPORT, I get a `Failed to find xref - java.lang.IllegalArgumentException: Failed to locate essentual files that belong to the site`.  It appears there is no `overview-summary.html`, `overview-frame.html`, or `index.html` in the `xref` directory.  There is a `index.html` and `project-summary.html`, etc. in the base sites dir.  Am I doing something wrong?  Thanks, Jon.

romani commented:
please share with me all commands you run and all possible logs (you can use github gist).

jonmbake commented:
@romani Here is the steps I took: https://gist.github.com/jonmbake/2eaa48f71b5f95a12ba3.  Thanks for your help.

romani commented:
> Ran laucher.sh. Had to modify it slightly to use gsed

Please descride a reason of such update. What OS you are on ? We need to make this tool to work for most OS, so you are welcome with patch.

> The script appears to be failing b/c there are missing files in xref

script parse report by `sed` to get list of files that referenced in a report and store them in file. It removes all files that are not referenced in this list of files.
If parsing is failed for any reason, it will remove all files, so it might be a problem for `ashm` tool.

Please let me know amount of "java" files at ~/git/checkstyle-contribution/checkstyle-tester/site_forked/xref/

It might be reasonable to update a script to recheck that if report has violations, at least one java file has to be at xref folder.

jonmbake commented:
> Ran laucher.sh. Had to modify it slightly to use gsed

I am running it on a Mac OS 10.10.  When I tried to run it with Mac's (BSD's) version of sed, I get the error:

```
sed: illegal option -- r
usage: sed script [-Ealn] [-i extension] [file ...]
       sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
```

> Please let me know amount of "java" files at ~/git/checkstyle-contribution/checkstyle-tester/site_forked/xref/

It looks like there are none.

```
> find ~/git/checkstyle-contribution/checkstyle-tester/site_forked/xref/ -type f -name '*.java' | wc -l
       0
```

Thanks again for helping me out.

romani commented:
You are welcome to update a script to support MacOS. Let me know if you do not have time to do this , I will try to find a solution. I do not have Mac.

jonmbake commented:
It's possible to install GNU sed (the version that comes the most linux distros) on Mac through brew, which I did.  I don't understand why the script failed when I updated it to use GNU's version of sed.  Which version of sed does the script work with?

romani commented:
I have, ubuntu 12.04:
```
~ $ sed --version
GNU sed version 4.2.1
Copyright (C) 2009 Free Software Foundation, Inc.
```

romani commented:
@jonmbake , fyi https://github.com/checkstyle/contribution/pull/48

romani commented:
@jonmbake , sorry for misleading, under xref there should be "html" files not a "java"
```
~/java/git-others/checkstyle/contribution/checkstyle-tester/target/site/xref [master|✔] $ find . -type f -name "*.html" | wc -l
254
```

jonmbake commented:
@romani I think the problem is that `overview-summary.html`, `overview-frame.html` and `allclasses-frame.html` is absent in `xref` and `ashm` expects them to be there.

I can try editing `XREF_FILES` in `attatrol.ahsm.site.InitialSiteVerifier`: https://github.com/attatrol/ahsm/blob/master/src/attatrol/ahsm/site/InitialSiteVerifier.java.

romani commented:
It is my bad (checkstyle-tester), xref cleanup should not remove them, I will update a script in few minutes

romani commented:
done.

please try again on latest code.

sabaka commented:
@romani, I have an error on latest code:
```bash
Backuping files that are refenced in report ...
/Users/idubinin/git/OS/contribution/checkstyle-tester/target/site/xref
mv: rename ./spring-framework/spring-aop/src/main/java/org/springframework/aop/framework/adapter/package-info.html to ./spring-framework/spring-aop/src/main/java/org/springframework/aop/framework/adapter/package-info.html.save: No such file or directory
```
I use only spring project to test on: https://github.com/sabaka/contribution/blob/own_config/checkstyle-tester/projects-to-test-on.properties
And links to code from report are dead.

romani commented:
@sabaka ,
it works fine for Linux-es - https://travis-ci.org/checkstyle/checkstyle/jobs/98600874#L474
you know I do not have MAC, I can not help you without sharing a screen with me.
Shell is pretty simple you can update it yourself to make it work. If you need help call me and share a screen.
Link to shell lines - https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/launch.sh#L107

romani commented:
I started build system for this project to resolve this problem ones for forever - https://travis-ci.org/checkstyle/contribution/builds

jonmbake commented:
@romani I updated the pull request with your suggestions.  Namely I:

1) Added additional UT cases
2) Moved RESOURCES token type check out of `#isMultilineStatement`

Here is the [Diff Report}(http://jonmbake.github.io/ahsm_checkstyle/checkstyle_merged.html).

romani commented:
please squash 2 commits in one.

report looks good.

jonmbake commented:
@romani Rebased commits.  Let me know if there is anything else.

romani commented:
merged as FF.

Thanks a lot for your contribution.
